### PR TITLE
GAWB-2933 Add LDAP schema for proxy email and email naming function

### PR DESCRIPTION
**Pull request by breilly2**
This goes hand-in-hand with https://github.com/broadinstitute/firecloud-develop/pull/949. I was trying to keep the small changes isolated so they could be implemented, reviewed, and merged independently. I apologize that I accidentally have 2 pieces here. The group email function in the migration script goes along with the one here but then also makes the updates to LDAP which required the schema work to be done.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/schema/JndiSchemaDAO.scala [hunk 1]
```diff
@@ -78,9 +81,34 @@ class JndiSchemaDAO(protected val directoryConfig: DirectoryConfig)(implicit exe
...
       _ <- removeWorkbenchGroupSchema()
       _ <- removeWorkbenchPetServiceAccountSchema()
+      _ <- removeWorkbenchPersonSchema()
     } yield ()
   }
 
+  private def createWorkbenchPersonSchema(): Future[Unit] = withContext { ctx =>
+    val schema = ctx.getSchema("")
+
+    createAttributeDefinition(schema, "1.3.6.1.4.1.18060.0.4.3.2.30", Attr.proxyEmail, "proxy group email", true)
```
Comment (breilly2) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/schema/JndiSchemaDAO.scala:91]: The IDs here are unique within sam and uses the same base but the final number was chosen arbitrarily. This is all mainly copying similar examples, primarily for pet service accounts. There's a good chance that there will be _something_ that needs tweaking here.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala [hunk 2]
```diff
@@ -27,8 +27,22 @@ object GoogleExtensions {
...
+  private val maxGroupEmailLength = 64
+
+  private[google] def toProxyFromUser(user: WorkbenchUser): WorkbenchEmail = {
+    val maxUsernameLength = maxGroupEmailLength -
+      "_".length -
+      user.id.value.length -
+      "@".length -
+      googleServicesConfig.appsDomain.length
+    val username = user.email.value.split("@").head
+    WorkbenchEmail(s"${username.take(maxUsernameLength)}_${user.id.value}@${googleServicesConfig.appsDomain}")
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala]: how about
```
val emailSuffix = s"_${user.id.value}@${googleServicesConfig.appsDomain}"
val username = user.email.value.split("@").head
val maxUsernameLength = maxGroupEmailLength - emailSuffix.length
WorkbenchEmail(username.take(maxUsernameLength) + emailSuffix)
```

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/schema/JndiSchemaDAO.scala (hunk 1 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/schema/JndiSchemaDAO.scala:91]: if they are unique within sam then 👍

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala [hunk 3]
```diff
@@ -304,6 +304,34 @@ class GoogleExtensionSpec(_system: ActorSystem) extends TestKit(_system) with Fl
...
+    val config = googleServicesConfig.copy(appsDomain = appsDomain)
+    val googleExtensions = new GoogleExtensions(null, null, null, null, null, null, null, config, null, null)
+
+    val user = WorkbenchUser(WorkbenchUserId(subjectId), WorkbenchEmail(s"$username@test.org"))
+
+    val proxyEmail = googleExtensions.toProxyFromUser(user).value
+    proxyEmail shouldBe "foo_0123456789@test.cloudfire.org"
+    proxyEmail should include (username)
+    proxyEmail should include (subjectId)
+    proxyEmail should include (appsDomain)
```
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala:321]: seems redundant to check "proxyEmail shouldBe ..." and all the should includes

**Review by dvoet — APPROVED** [review]

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala (hunk 3 above)
Comment (breilly2) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala:321]: It technically is. However the requirement is (and the description of the test case indicates) that each of those items is in the email address. Having the elements as separate strings along with the individual asserts makes it clear that those facts are true, more so than having to mentally match the overall `shouldBe`.

An alternative would be simply:
```
proxyEmail shouldBe s"${username}_$subjectId@$appsDomain"
```
which feels a little too much like a re-implementation of the function but is probably okay.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala (hunk 2 above)
Comment (breilly2) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala]: I like that this builds in the "_" and "@" into the length calculation automatically. I'll incorporate that. Thanks.

**Review by MatthewBemis — APPROVED** [review]

**Comment by breilly2** [comment]
jenkins retest because fiab-start failed

**Pull request merged**